src/H5E.c

```diff
@@ -1217,22 +552,30 @@ H5Epush2(hid_t err_stack, const char *file, const char *func, unsigned line, hid
     va_start(ap, fmt);
     va_started = true;
 
-    /* Use the vasprintf() routine, since it does what we're trying to do below */
-    if (HDvasprintf(&tmp, fmt, ap) < 0)
-        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
+    /* Duplicate string information */
+    if (NULL == (tmp_file = strdup(file)))
```

Comment:
Duplicate strings and increment IDs for application error push operations.   Do this here, in the API routine, instead of within the library.